### PR TITLE
#180 add confirms

### PR DESCRIPTION
**Pull request by je-ik**

### CODE REVIEW

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 1]
```diff
@@ -186,7 +194,7 @@ public boolean onNext(StreamElement element, BatchLogObserver.OnNextContext cont
     return enqueue(element, new BatchLogObserverUnifiedContext(context));
   }
 
-  private boolean enqueue(StreamElement element, @Nullable UnifiedContext context) {
+  private boolean enqueue(StreamElement element, UnifiedContext context) {
     try {
       Preconditions.checkArgument(element == null || context != null);
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:199]: should this be `element != null && context != null`? (according to new method signature)

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 2]
```diff
@@ -143,10 +149,12 @@ public Offset getOffset() {
   private final AtomicLong watermark;
   private final BlockingQueue<Pair<StreamElement, UnifiedContext>> queue;
   private final AtomicBoolean stopped = new AtomicBoolean();
+  private volatile boolean nackAllIncoming = false;
   @Getter @Nullable private UnifiedContext lastWrittenContext;
   @Getter @Nullable private UnifiedContext lastReadContext;
   private long limit;
   private boolean cancelled = false;
+  private transient CountDownLatch cancelledLatch = new CountDownLatch(1);
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:157]: unnecessary `cancelled` flag, can be replaced with `cancelledLatch.getCount() > 0`

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 3]
```diff
@@ -302,11 +317,16 @@ void stop() {
...
+    nackAllIncoming = nack;
     stopped.set(true);
     if (nack) {
       List<Pair<StreamElement, UnifiedContext>> drop = new ArrayList<>();
       queue.drainTo(drop);
       drop.stream().map(Pair::getSecond).filter(Objects::nonNull).forEach(UnifiedContext::nack);
+      Optional.ofNullable(lastWrittenContext).ifPresent(UnifiedContext::nack);
+    }
+    if (getWatermark() < Watermarks.MAX_WATERMARK) {
+      ExceptionUtils.ignoringInterrupted(() -> cancelledLatch.await(1, TimeUnit.SECONDS));
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:329]: what are the consequences of latch timing out?

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 4]
```diff
@@ -302,11 +317,16 @@ void stop() {
   }
 
   void stop(boolean nack) {
+    nackAllIncoming = nack;
     stopped.set(true);
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:321]: this is confusing, `stopped` flag is atomic and `nackAllIcoming` is not 🤔 why is the reason for this?

Comment (dmvk) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:321]: after closer look, stooped can be replaced with `volatile` as we're not using any atomic methods (eg. `getAndSet()`)

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 5]
```diff
@@ -223,12 +232,18 @@ private boolean putToQueue(@Nullable StreamElement element, @Nullable UnifiedCon
...
+      if (nackAllIncoming && context != null) {
+        context.nack();
+      }
+    } else {
+      while (!stopped.get()) {
+        if (queue.offer(p, 50, TimeUnit.MILLISECONDS)) {
+          return true;
+        }
       }
+      log.debug("{}: Finishing consumption due to source being stopped", name);
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:245]: If I understand this correctly:
- Stop can be called asynchronously
- If source is stopped, we want to nack current element

Nacking seems broken, in case `stop` signal is received inside `while` loop

Should the code look like this instead?
```java
      while (!stopped.get()) {
        if (queue.offer(p, 50, TimeUnit.MILLISECONDS)) {
          return true;
        }
      }
      if (nackAllIncoming && context != null) {
        return true;
        context.nack();
      }
      log.debug("{}: Finishing consumption due to source being stopped", name);
      return false;
```

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java [hunk 6]
```diff
@@ -302,11 +317,16 @@ void stop() {
   }
 
   void stop(boolean nack) {
+    nackAllIncoming = nack;
     stopped.set(true);
     if (nack) {
       List<Pair<StreamElement, UnifiedContext>> drop = new ArrayList<>();
       queue.drainTo(drop);
       drop.stream().map(Pair::getSecond).filter(Objects::nonNull).forEach(UnifiedContext::nack);
+      Optional.ofNullable(lastWrittenContext).ifPresent(UnifiedContext::nack);
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:326]: isn't nacking handled in that busy waiting loop above (putToQueue)?

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java [hunk 7]
```diff
@@ -112,10 +131,17 @@ public ProcessContinuation process(
             return ProcessContinuation.stop();
           }
           output.outputWithTimestamp(element, Instant.ofEpochMilli(element.getStamp()));
+          readContext.set(observer.getLastReadContext());
         }
       } finally {
-        observer.stop(false);
+        // nack all buffered but unprocessed data iff we have not emitted any output
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java:137]: typo: `iff`

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java [hunk 8]
```diff
@@ -166,6 +192,7 @@ public boolean onError(Throwable error) {
 
         @Override
         public boolean onNext(StreamElement ingest, OnNextContext context) {
+          context.nack();
```
Comment (dmvk) [inline beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java:195]: why is this needed?

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLogTest.java [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.junit.Test;
+
+/** Test {@link ListCommitLog}. */
+public class ListCommitLogTest {
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLogTest.java:42]: 👍

File: beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java [hunk 10]
```diff
@@ -54,26 +54,36 @@
 
   @Test(timeout = 30000)
   public void testReadingFromCommitLog() {
```
Comment (dmvk) [inline beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java:56]: how is the new functionality (confirms) stress by these tests? I probably I'm missing something obvious here

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java (hunk 7 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java:137]: `iff` if abbrebiaviation of `if and only if`.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 1 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:199]: Yes, I think I fixed this one PR later, but can do it here.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 2 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:157]: That's technically correct, but the flag is no overhead and makes things a little more explicit. I'd like to keep it there.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 4 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:321]: Yes, we can probably do that.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 6 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:326]: Actually no. There are quite many states and code-paths, we must make sure that we nack all read-but-not-consumed data. `putToQueue` might not be called at all or it might be called before call to `stop()`, there are many corner cases.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java (hunk 8 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java:195]: Because otherwise we would hold the consumed element until timeout (or forever in case of `ListCommitLog`, which doesn't timeout elements).

File: beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java (hunk 10 above)
Comment (je-ik) [reply beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java:56]: here:
```java
ListCommitLog commitLog = ListCommitLog.ofNonExternalizable(data, direct.getContext());
```
That creates the new `ListCommitLog`, which mimics non-externalizable offsets (that are offsets that cannot be persisted in external system - checkpoint - and be sure they represent the same element even upon recovery). Commit-logs with "non-externalizable offsets" (e.g. Google PubSub) rely heavily on ack() and nack() of elements to ensure at-least-once semantics (typically not exactly once-semantics, because when consumer consumes element and does neither ack() nor nack() it until timeout, the element is resend to another (or the same) consumer).

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 2 above)
Comment (dmvk) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:157]: one think is that the flag is not transient and CountDownLatch is not, so it may not be consistent after serde cycle

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java (hunk 8 above)
Comment (dmvk) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java:195]: 👍 makes sense, thanks

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 2 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:157]: The latch is not carrying semantic information. It is just a synchronization primitive (which is why it is transient). A cancelled observer remains cancelled even after (de)serialization which is also consistent.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 3 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:329]: Hm, interesting question. It might result in element not being correctly `nack`ed, which would in most implementations probably result in that this element would be retried after some timeout, but will be probably late. The other option of not setting a timeout for the latch might result (probably under the same circumstances as the the timeout in the other case) would result in pipeline being stuck, which seems to be worse case IMO.

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 5 above)
Comment (je-ik) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:245]: Hm, nice catch. That is probably the reason why `lastWrittenContext` is explicitly nacked in the `stop` method. I'll try this. :+1:

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java (hunk 6 above)
Comment (dmvk) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/BlockingQueueLogObserver.java:326]: related to https://github.com/datadrivencz/proxima-platform/pull/460#discussion_r533207074

File: beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java (hunk 7 above)
Comment (dmvk) [reply beam/core/src/main/java/cz/o2/proxima/beam/direct/io/CommitLogRead.java:137]: hmm, ok :) I'd maybe better not to use these, but I'll leave this up to you

File: beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java (hunk 10 above)
Comment (dmvk) [reply beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java:56]: thanks for the explanation ;) This sheds some light to ListCommitLog internals 👍

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLog.java [hunk 11]
```diff
@@ -93,35 +113,175 @@ public static ListCommitLog of(List<StreamElement> data, Context context) {
...
+
+  @VisibleForTesting
+  final class ListObserveHandle implements ObserveHandle {
 
-  private static final class NopObserveHandle implements ObserveHandle {
+    @Getter private final String consumerName;
+
+    @Getter private volatile boolean closed = false;
+
+    ListObserveHandle(String logUuid, String consumerName) {
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLog.java:127]: unused `logUuid`

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLog.java [hunk 12]
```diff
@@ -93,35 +113,175 @@ public static ListCommitLog of(List<StreamElement> data, Context context) {
...
 
-  private final List<StreamElement> data;
-  private final Context context;
-  private transient ExecutorService executor;
+  public static ListCommitLog ofNonExternalizable(List<StreamElement> data, Context context) {
+    return new ListCommitLog(data, context, false);
+  }
+
+  @VisibleForTesting
+  final class ListObserveHandle implements ObserveHandle {
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLog.java:121]: `static`

File: beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java (hunk 10 above)
Comment (dmvk) [reply beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java:56]: Can you please add this explation to `cz.o2.proxima.direct.storage.ListCommitLog#ofNonExternalizable` javadoc?

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLog.java [hunk 13]
```diff
@@ -93,35 +113,175 @@ public static ListCommitLog of(List<StreamElement> data, Context context) {
...
+            if (succ) {
+              ack(offset);
+            } else {
+              nack(offset);
+            }
+          };
+      if (bulk) {
+        contextCommitter =
+            (succ, exc) -> {
+              synchronized (inflightOffsets) {
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLog.java:242]: Why is the synchronization needed here?

File: direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLog.java [hunk 14]
```diff
@@ -93,35 +113,175 @@ public static ListCommitLog of(List<StreamElement> data, Context context) {
...
+                // clone to prevent ConcurrentModificationException
+                new ArrayList<>(inflightOffsets)
```
Comment (dmvk) [inline direct/core/src/test/java/cz/o2/proxima/direct/storage/ListCommitLog.java:244]: simplify:

```java
                new ArrayList<>(inflightOffsets)
                    .stream()
                    .filter(o -> o <= offset)
                    .map(offsetToContext::remove)
                    .filter(Objects::nonNull)
                    .forEach(p -> p.commit(succ, exc));
```

File: beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java (hunk 10 above)
Comment (je-ik) [reply beam/core/src/test/java/cz/o2/proxima/beam/direct/io/CommitLogReadTest.java:56]: ok

**Comment by je-ik** [comment]
@dmvk review incorporated

**Review by dmvk — APPROVED** [review]
LGTM 👍 Great job!
